### ISSUE TITLE
Destructors aren't run for detached tasks

### ISSUE THREAD

**Issue by vojtechkral** [smol-rs/smol#294]
I noticed today that when I `.detach()` a task, its `drop()` is never run.
Is this to be expected?
If yes, this should probably be documented in `Task::detach()` docs...

Using smol v1.3.0.

**Comment by notgull** [comment]
Tasks spawned with `smol::spawn` and detached are never dropped, this should be noted in that method's documentation. On the other hand, if you spawn tasks with a normal executor they should be dropped when the executor's destructor runs.

**Comment by vojtechkral** [comment]
It looks like when I _don't_ detach the task, the `drop()` is only run _sometimes_. I'm quite confused.

Looks like I have to `.cacnel().await`, then it runs the dtor for sure.

**Comment by notgull** [comment]
This is to be expected. The executor in this case is `'static` and static types don't always run their destructors.

The intended solution is to create your own executor.

**Comment by vojtechkral** [comment]
Ah, okay.

I can PR the docs change for `.detach()` if you'd like...

**Comment by notgull** [comment]
Yes, that would be appreciated.

**Comment by halvors** [comment]
> It looks like when I _don't_ detach the task, the `drop()` is only run _sometimes_. I'm quite confused.
> 
> Looks like I have to `.cacnel().await`, then it runs the dtor for sure.

I have an application starting detached tasks to handle each TCP connection, and this application seems to be leaking memory as a result of that, might this be the reason?

I am a little confused to how i should handle the destruction using a custom executor? Do you have any examples on how to do that properly?

**Comment by halvors** [comment]
@notgull Is it ok that i ask this here even thought the issue is closed?

**Comment by notgull** [comment]
> @notgull Is it ok that i ask this here even thought the issue is closed?

It's always okay to ask. I've just been distracted for the past week or so.

> I have an application starting detached tasks to handle each TCP connection, and this application seems to be leaking memory as a result of that, might this be the reason?

If you have an application like this it's probably already time to consider moving away from `smol::spawn`.

**Comment by vojtechkral** [comment]
> If you have an application like this it's probably already time to consider moving away from `smol::spawn`.

Excuse me, what should in such case be used instead?
